examples/pytorch/aml.py

```diff
@@ -0,0 +1,158 @@
+from ray.air import session
+from ray.air.config import ScalingConfig, RunConfig
+
+from deepgnn.graph_engine.data.cora import CoraFull
+from deepgnn.graph_engine.snark.distributed import Server, Client as DistributedClient
+
+
+@ray.remote
+class ServerLock:
+    """Lock for server."""
+
+    def __init__(self, size):
+        """Initialize server lock."""
+        self.cache = {}
+        self.locks = [0 for _ in range(size)]
+
+    def set_lock(self, i):
+        """Set lock i."""
+        self.locks[i] = 1
+
+    def release_lock(self, i):
+        """Release lock i."""
+        self.locks[i] = 2
+
+    def wait_set(self):
+        """See if all locks set."""
+        return all([v == 1 for v in self.locks])
+
+    def wait_released(self):
+        """See if all locks released."""
+        return all([v == 2 for v in self.locks])
+
+    def put(self, x, y):
+        """Put value y at x."""
+        self.cache[x] = y
+
+    def get(self, x):
+        """Get value x."""
+        return self.cache.get(x)
+
+
+class ServerContext:
+    """Server context."""
+
+    def __init__(self, world_size: int, rank: int, address: str):
+        """Initialize server context."""
+        self.world_size = world_size
+        self.rank = rank
+        self.address = address
+        self.server_lock = server_lock
+
+    def __enter__(self):
+        """Enter server context."""
+        self.server_lock.put.remote(self.rank, self.address)
+        self.server_lock.set_lock.remote(self.rank)
+        while not ray.get(self.server_lock.wait_set.remote()):
+            pass
+        return self
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        """Exit server context."""
+        self.server_lock.release_lock.remote(self.rank)
+        while not ray.get(self.server_lock.wait_released.remote()):
+            pass
+
+    def get_address(self):
+        """Return addresses of all servers connected."""
+        return ray.get([self.server_lock.get.remote(i) for i in range(self.world_size)])
+
+
+def train_func(config: dict):
+    """Training loop for ray trainer."""
+    cora = CoraFull(num_partitions=session.get_world_size())
+
+    hostname = (
+        "127.0.0.1"
+        if config["is_unit_test"]
+        else ray._private.services.get_node_ip_address()
+    )
+    address = f"{hostname}:999{session.get_world_rank()}"
+
+    Server(address, cora.data_dir(), session.get_world_size(), 1)
+
+    with ServerContext(
+        session.get_world_size(), session.get_world_rank(), address
+    ) as lock:
+        cl = DistributedClient(lock.get_address())
+
+        features = cl.node_features(np.array([0, 1]), np.array([[0, 1]]), np.float32)
+        npt.assert_equal(features, np.array([[0.0], [0.0]]))
+
+
+if __name__ == "__main__":
+    import sys
+    from azureml.core import Workspace
+    from ray_on_aml.core import Ray_On_AML
+
+    aml = True
+    if len(sys.argv) < 2 or sys.argv[1] != "--unit_test":
+        ws = Workspace.from_config()
```

Comment:
why do we need lines 116-135 in this file?

Reply:
This is so it will actually run for users in multi node AML

Reply:
How do we know it will keep working if it is never executed in CI?

Reply:
It is simple code from ray_on_aml which is heavily tested by them, so we can count on it